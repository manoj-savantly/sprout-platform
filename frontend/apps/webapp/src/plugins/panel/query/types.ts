import { QueryParametersConfiguration } from './editors/QueryParametersEditor';
import { CodeTemplateEditorConfiguration } from './editors/TemplateEditor';

export type QueryPanelDisplayType = 'FRAME' | 'RENDER';

export interface QueryPanelOptions {
  url: string;
  queryParameters: QueryParametersConfiguration;
  displayType: QueryPanelDisplayType;
  template: CodeTemplateEditorConfiguration;
  useTemplate: boolean;
  useProxy: boolean;
  headers: string;
}
