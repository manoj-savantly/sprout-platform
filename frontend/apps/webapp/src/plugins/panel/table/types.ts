import { QueryParametersConfiguration } from './editors/QueryParametersEditor';
import { CodeTemplateEditorConfiguration } from './editors/TemplateEditor';

export interface TablePanelOptions {
  url: string;
  queryParameters: QueryParametersConfiguration;
  jsonDataPath: string;
  headerTemplate: CodeTemplateEditorConfiguration;
  useProxy: boolean;
  headers: string;
}
