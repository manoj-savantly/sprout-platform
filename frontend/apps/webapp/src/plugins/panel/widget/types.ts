export interface WidgetPanelOptions {
  dataSourceId: string
  dataId: string
}

export type WidgetDataType = 'MARKDOWN' | 'MARKUP' | 'JSON'

export interface WidgetData {
    id: string;
    name: string;
    dataType: WidgetDataType;
    data: any;
}
