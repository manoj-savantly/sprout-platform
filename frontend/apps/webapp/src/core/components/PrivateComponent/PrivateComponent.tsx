import React, { ReactNode, useMemo } from 'react';
import { useSelector } from 'react-redux';
import { StoreState } from '../../../types';

const authorize = (authorities: string[], hasAnyAuthorities?: string[]) => {
  if (!hasAnyAuthorities || hasAnyAuthorities.length === 0) {
    return true;
  } else {
    if (!authorities && hasAnyAuthorities.length !== 0) {
      return false;
    } else {
      return hasAnyAuthorities.some((auth) => authorities.includes(auth));
    }
  }
};

export const PrivateComponent = ({
  hasAnyAuthority,
  children
}: {
  hasAnyAuthority?: string[];
  children: ReactNode;
}) => {
  const sessionHasBeenFetched = useSelector((state: StoreState) => state.authentication.sessionHasBeenFetched);
  const user = useSelector((state: StoreState) => state.authentication.user);
  const authorized = useMemo(() => {
    return sessionHasBeenFetched && authorize(user.authorities, hasAnyAuthority);
  }, [sessionHasBeenFetched, user]);

  return <React.Fragment>{authorized && children}</React.Fragment>;
};
