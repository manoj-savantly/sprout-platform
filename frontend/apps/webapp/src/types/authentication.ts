import { User } from '@savantly/sprout-api';

export interface AuthenticationState {
  loading: boolean;
  isAuthenticated: boolean;
  loginError: boolean; // Errors returned from server side
  errorMessage?: string; // Errors returned from server side
  sessionHasBeenFetched: boolean;
  sessionFetchFailed: boolean;
  logoutUrl?: string;
  user: User;
  showLogin: boolean;
}

export interface AuthenticationUpdate {
  errorMessage?: string;
  accessToken?: string;
}

export interface OAuthClientConfig {
  name: string;
  autoLogin?: boolean;
  displayName: string;
  issuerUri: string;
  clientId: string;
  scope: string;
  authorizationUrl: string;
  redirectUrl: string;
  tokenUrl: string;
  userInfoUrl: string;
}