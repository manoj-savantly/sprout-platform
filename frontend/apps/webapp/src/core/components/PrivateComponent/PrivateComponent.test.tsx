import React from 'react';
import { Provider } from 'react-redux';
import { act, create, ReactTestRenderer } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { AuthenticationState } from '../../../types';
import { PrivateComponent } from './PrivateComponent';

const mockAuthenticationState: AuthenticationState = {
  isAuthenticated: true,
  loading: false,
  loginError: false,
  sessionFetchFailed: false,
  sessionHasBeenFetched: true,
  showLogin: false,
  user: {
    authorities: ['EVERYONE'],
    name: 'user'
  }
};

const mockStore = configureStore([]);

describe('PrivateComponent', () => {
  let component: ReactTestRenderer;
  //let container: HTMLElement;
  let store: any;
  beforeEach(() => {
    // setup a DOM element as a render target
    //container = document.createElement('div');
    //document.body.appendChild(container);
    store = mockStore({ authentication: mockAuthenticationState });
  });

  afterEach(() => {
    // cleanup on exiting
    //container && unmountComponentAtNode(container) && container.remove();
    //container = null;
  });

  it('renders when there is no authority requirement', () => {
    act(() => {
      component = create(
        <Provider store={store}>
          <PrivateComponent hasAnyAuthority={[]}>true</PrivateComponent>
        </Provider>
      );
    });
    const ss = component.toJSON();
    expect(ss).toBe('true');
    expect(ss).toMatchSnapshot();
  });

  it('doesnt render when the user lacks authority', () => {
    act(() => {
      component = create(
        <Provider store={store}>
          <PrivateComponent hasAnyAuthority={['ADMIN', 'FM_ADMIN']}>true</PrivateComponent>
        </Provider>
      );
    });
    const ss = component.toJSON();
    expect(ss).not.toBe('true');
    expect(ss).toMatchSnapshot();
  });

  it('renders when the user has matching authority', () => {
    act(() => {
      component = create(
        <Provider store={store}>
          <PrivateComponent hasAnyAuthority={['EVERYONE', 'FM_ADMIN']}>true</PrivateComponent>
        </Provider>
      );
    });
    const ss = component.toJSON();
    expect(ss).toBe('true');
    expect(ss).toMatchSnapshot();
  });
});
