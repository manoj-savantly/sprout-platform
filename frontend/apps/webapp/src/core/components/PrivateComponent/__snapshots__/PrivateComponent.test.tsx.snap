// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateComponent doesnt render when the user lacks authority 1`] = `null`;

exports[`PrivateComponent renders when the user has matching authority 1`] = `"true"`;

exports[`PrivateComponent renders when there is no authority requirement 1`] = `"true"`;
